File: server/src/main/kotlin/net/contrapt/jvmcode/service/ParseService.kt
package net.contrapt.jvmcode.service

import io.vertx.core.logging.LoggerFactory
import net.contrapt.jvmcode.model.*
import java.io.File
import java.lang.IllegalArgumentException

class ParseService(val symbolRepository: SymbolRepository) {

    private val logger = LoggerFactory.getLogger(javaClass)

    fun parse(request: ParseRequest, parser: LanguageParser?) : ParseResult {
        if ( parser == null ) throw IllegalStateException("No parser found for ${request.languageId}")
        val result = parser.parse(request)
        resolveSymbols(result)
        return result
    }

    fun resolveSymbols(result: ParseResult) {
        val pkgName = result.pkg?.name ?: ""
        val importMap = result.imports.associate { it.name.substringAfterLast(".") to it.name }
        result.symbols.forEach {
            when (it.symbolType) {
                ParseSymbolType.TYPEREF -> it.type = typeRefName(it, importMap, pkgName)
                ParseSymbolType.FIELD -> it.type = typeRefName(it,importMap, pkgName)
                ParseSymbolType.VARIABLE ->  it.type = typeRefName(it, importMap, pkgName)
                ParseSymbolType.METHOD -> it.type = typeRefName(it, importMap, pkgName)
                ParseSymbolType.CLASS,
                ParseSymbolType.INTERFACE,
                ParseSymbolType.ENUM,
                ParseSymbolType.THIS,
                ParseSymbolType.OBJECT-> it.type = "$pkgName.${typeDefName(it, result.symbols)}"
                ParseSymbolType.CONSTRUCTOR -> it.type = "$pkgName.${constructorType(it, result.symbols)}"
                else -> {}
            }
        }
        result.symbols.forEach {
            when (it.symbolType) {
                ParseSymbolType.SYMREF -> resolveSymbolRef(it, result.symbols)
                else -> {}
            }
            if (it.symbolDef == null) it.type = typeRefName(it, importMap, pkgName)
        }
    }

    private fun typeRefName(symbol: ParseSymbol, importMap: Map<String, String>, pkg: String) : String? {
        // If it is imported, use that
        val imported = importMap[symbol.type]
        if ( imported != null) return imported;
        // Otherwise, look for all the unqualified names that match
        val names = symbolRepository.getJarEntriesByName(symbol.type ?: "")
        val autoType = if (names.size == 1) names[0] else null // Is it auto-imported?
        val wildTypes = names.filter { importMap.values.contains("${it.pkg}.*") }
        val packageType = names.firstOrNull { it.pkg == pkg } // Is it in the same package?
        return when {
            autoType != null -> autoType.fqcn
            wildTypes.isNotEmpty() -> wildTypes.first().fqcn
            packageType != null -> packageType.fqcn //Not until we are loading the local classes
            else -> symbol.type // could default to same package here also
        }
    }

    private fun constructorType(symbol: ParseSymbol, symbols: List<ParseSymbol>) : String {
        val type = if (symbol.parent > -1) symbols[symbol.parent] else symbol
        return typeDefName(type, symbols)
    }

    private fun typeDefName(symbol: ParseSymbol, symbols: List<ParseSymbol>) : String {
        if (symbol.parent == -1) return symbol.name
        else return "${typeDefName(symbols[symbol.parent], symbols)}.${symbol.name}"
    }

    private fun resolveSymbolRef(symbol: ParseSymbol, symbols: MutableList<ParseSymbol>) {
        val caller = when (val c = symbol.caller) {
            null -> null
            else -> symbols[c]
        }
        when (caller?.name) {
            null -> symbol.symbolDef = findSymbolDef(symbol, symbol.parent, symbols)
            "this" -> symbol.symbolDef = findSymbolDef(symbol, symbol.parent, symbols, true)
            // TODO super
        }
        when (val sd = symbol.symbolDef) {
            null -> {}
            else -> symbol.type = symbols[sd].type
        }
    }

    private fun findSymbolDef(symbol: ParseSymbol, parent: Int, symbols: List<ParseSymbol>, isThis: Boolean = false) : Int? {
        if (parent == -1) return null
        if (parent == symbol.id) return null
        val def = symbols[parent].children.reversed().firstOrNull { id ->
            val child = symbols[id]
            child.name == symbol.name &&
                ((isThis && symbol.symbolType.isMember) || (!isThis && child.symbolType.isDef))
        }
        if (def == null) return findSymbolDef(symbol, symbols[parent].parent, symbols)
        else return def
    }

}

File: src/parse_service.ts
import { ParseRequest, ParseResult, LanguageRequest, ParseSymbolType, ParseSymbol } from "./language_model";
import { ProjectRepository } from "./project_repository";

export class ParseService {

    private repository: ProjectRepository

    constructor(repository: ProjectRepository) {
        this.repository = repository
    }

    public async parse(request: ParseRequest, language: LanguageRequest) : Promise<ParseResult> {
        let result = await language.parse(request)
        this.resolveSymbols(result, language)
        return result
    }

    /**
     * Resolve symbol types as much as possible
     * @param result 
     * @param language 
     */
    public resolveSymbols(result: ParseResult, language: LanguageRequest) {
        let pkgName = result.pkg ? result.pkg.name ? result.pkg.name : "" : ""
        let importMap = new Map<string, string>()
        language.imports.concat(result.imports.map(i=>i.name)).forEach(imp => {
            let key = imp.split('.').slice(-1)[0]
            key = key === '*' ? imp : key
            importMap.set(key, imp)
        })
        importMap.set(pkgName, pkgName)
        result.symbols.forEach(sym => {
            switch(sym.symbolType) {
                case ParseSymbolType.TYPEREF:
                case ParseSymbolType.FIELD:
                case ParseSymbolType.VARIABLE:
                case ParseSymbolType.METHOD: sym.type = this.typeRefName(sym, importMap, pkgName); break;
                case ParseSymbolType.CLASS:
                case ParseSymbolType.INTERFACE:
                case ParseSymbolType.ENUM:
                case ParseSymbolType.THIS:
                case ParseSymbolType.OBJECT: sym.type = `${pkgName}.${this.typeDefName(sym, result.symbols)}`; break;
                case ParseSymbolType.CONSTRUCTOR: sym.type = `${pkgName}.${this.constructorType(sym, result.symbols)}`; break;
                default: {}
            }
        })
        result.symbols.forEach(sym =>{
            if (sym.symbolType === ParseSymbolType.SYMREF) {
                this.resolveSymRef(sym, result.symbols)
            }
            if (sym.symbolDef == null) sym.type = this.typeRefName(sym, importMap, pkgName)
        })
    }

    private typeRefName(sym: ParseSymbol, importMap: Map<string, string>, pkg: string) : string {
        // If it is imported, use that
        let imported = importMap.get(sym.type)
        if (imported) return imported;
        let imports = Array.from(importMap.values())
        // Look for a local class in same package
        let classes = this.repository.findClassDataByName(sym.type ? sym.type : '')
        let localClass = classes.find(c => c.pkg === pkg)
        if (localClass) return localClass.fqcn
        // Otherwise, look for all the unqualified names that match
        let entries = this.repository.findJarEntriesByName(sym.type ? sym.type : '')
        // Is it auto imported?
        let autoType = entries.find(entry => imports.includes(entry.pkg))
        if (autoType) return autoType.fqcn
        // Is it wildcard imported?
        let wildType = entries.find(entry => imports.includes(`${entry.pkg}.*`))
        if (wildType) return wildType.fqcn
        // Is it in the same package (but from a jar)
        let packageType = entries.find(entry => entry.pkg === pkg)
        if (packageType) return packageType.fqcn
        // Otherwise, leave as is
        return sym.type
    }

    private typeDefName(sym: ParseSymbol, symbols: ParseSymbol[]) : string {
        if (sym.parent == -1) return sym.name
        else {
            let parentName = this.typeDefName(symbols[sym.parent], symbols)
            return `"${parentName}.${sym.name}`
        }
    }

    private constructorType(sym: ParseSymbol, symbols: ParseSymbol[]) : string {
        let type = (sym.parent > -1) ? symbols[sym.parent] : sym
        return this.typeDefName(type, symbols)
    }

    private resolveSymRef(sym: ParseSymbol, symbols: ParseSymbol[]) {
        let caller = (sym.caller) ? symbols[sym.caller] : undefined
        sym.symbolDef = caller ? (caller.name === 'this' ? this.findSymbolDef(sym, sym.parent, symbols, true) : undefined) : this.findSymbolDef(sym, sym.parent, symbols)
        sym.type = sym.symbolDef ? symbols[sym.symbolDef].type : undefined
    }

    private findSymbolDef(sym: ParseSymbol, parent: number, symbols: ParseSymbol[], isThis?: boolean) : number {
        if (parent == -1) return undefined
        if (parent == sym.id) return undefined
        let def = symbols[parent].children.reverse().find(id => {
            let child = symbols[id]
            return child.name == sym.name &&
                ((isThis && this.isMember(sym.symbolType)) || (!isThis && this.isDef(child.symbolType)))
        })
        if (!def) return this.findSymbolDef(sym, symbols[parent].parent, symbols)
        else return def
    }

    private isMember(type: ParseSymbolType) : boolean {
        return [
            ParseSymbolType.INTERFACE,
            ParseSymbolType.ANNOTATION,
            ParseSymbolType.ENUM,
            ParseSymbolType.OBJECT,
            ParseSymbolType.CONSTRUCTOR,
            ParseSymbolType.METHOD,
            ParseSymbolType.FIELD,
            ParseSymbolType.VARIABLE
        ].includes(type)
    }

    private isDef(type: ParseSymbolType) : boolean {
        return [ParseSymbolType.METHOD, ParseSymbolType.FIELD, ParseSymbolType.VARIABLE].includes(type)
    }
}